src/twister2/scripts/hardware_map.py

```diff
@@ -48,108 +55,145 @@
-                runner='unknown',
-                connected=True
-            )
-
-            for runner in RUNNER_MAPPING.keys():
-                products = RUNNER_MAPPING.get(runner)
+        if device.manufacturer not in MANUFACTURER:
+            logger.warning('Unsupported device (%s): %s' % (device.manufacturer, device))
+            continue
+
+        # TI XDS110 can have multiple serial devices for a single board
+        # assume endpoint 0 is the serial, skip all others
+        if (
+            device.manufacturer == 'Texas Instruments'
+            and device.location is not None
+            and not device.location.endswith('0')
+        ):
+            continue
+
+        hardware_map: HardwareMap = HardwareMap(
+            platform='unknown',
+            id=device.serial_number or '',
+            serial=persistent_map.get(device.device, device.device),
+            product=device.product or '',
+            runner='unknown',
+            connected=True
+        )
+
+        # find runner for the device
+        if device.product is not None:
+            for runner, products in RUNNER_MAPPING.items():
                 if device.product in products:
                     hardware_map.runner = runner
                     continue
                 # Try regex matching
-                for product in products:
-                    if re.match(product, device.product):
-                        hardware_map.runner = runner
-
-            hardware_map.connected = True
-            hardware_map.lock = None
-            hardware_map_list.append(hardware_map)
-        else:
-            logger.warning('Unsupported device (%s): %s' % (device.manufacturer, device))
+                if any(re.match(product, device.product) for product in products):
+                    hardware_map.runner = runner
+
+        hardware_map_list.append(hardware_map)
 
     return hardware_map_list
 
 
 def write_to_file(filename: str | Path, hardware_map_list: list[HardwareMap]) -> None:
     """Save hardware map to file."""
-    new_hardware_map = hardware_map_list.copy()
+    new_hardware_map_list = hardware_map_list
+    merged_hardware_map_list = []
 
     # update existing hardware map
     if os.path.exists(filename):
         old_hardware_map_list = HardwareMap.read_from_file(filename)
         for hardware in old_hardware_map_list:
             hardware.connected = False
-            hardware.serial = None
+            hardware.serial = ''
 
-        while new_hardware_map:
-            new_hardware = new_hardware_map.pop()
-            for old_hardware in reversed(old_hardware_map_list):
+        for new_hardware in new_hardware_map_list:
+            for old_hardware in old_hardware_map_list:
                 if old_hardware.id == new_hardware.id and old_hardware.product == new_hardware.product:
                     old_hardware.serial = new_hardware.serial
                     old_hardware.connected = True
                     break
```

Comment:
if you generate hardware map for nrf5340dk_nrf5340_cpuapp, you will receive 3 ports:
```
| platform   |           id | serial       |
|------------|--------------|--------------|
| unknown    | 000960185465 | /dev/ttyACM4 |
| unknown    | 000960185465 | /dev/ttyACM3 |
| unknown    | 000960185465 | /dev/ttyACM2 |
```
for nrf52840dk_nrf52840 there are two ports.

Then if you update map files with proper platform and runner, and run scripts once again, then you keep only info about first matched port:
```
- available: false
  baud: '115200'
  connected: true
  id: 000960185465
  platform: nrf5340dk_nrf5340_cpuapp
  product: J-Link
  runner: nrfjprog
  serial: /dev/ttyACM2
  ...
- available: false
  baud: '115200'
  connected: false
  id: 000960185465
  platform: nrf5340dk_nrf5340_cpuapp
  product: J-Link
  ...
```